components/mediation/data-publishers/org.wso2.micro.integrator.observability/src/main/java/org/wso2/micro/integrator/obsrvability/handler/metrics/publisher/MetricHandler.java

```diff
@@ -0,0 +1,480 @@
+    private void startTimers(MessageContext synCtx, String serviceName, String serviceType, String apiInvocationUrl) {
+
+        if (null == synCtx.getProperty("HAS_EXECUTED_FLOW")) {
+            switch (serviceType) {
+                case SynapseConstants.PROXY_SERVICE_TYPE:
+                    Map<String, String[]> proxyMap = new HashMap();
+                    proxyMap.put(MetricConstants.PROXY_LATENCY_SECONDS, new String[]{serviceName, serviceType});
+
+                    synCtx.setProperty(MetricConstants.PROXY_LATENCY_TIMER,
+                            metricReporterInstance.getTimer(MetricConstants.PROXY_LATENCY_SECONDS, proxyMap));
+                    break;
+                case MetricConstants.INBOUND_ENDPOINT:
+                    Map<String, String[]> inboundEndpointMap = new HashMap();
+                    inboundEndpointMap.put(MetricConstants.INBOUND_ENDPOINT_LATENCY_SECONDS, new String[]{serviceName,
+                            serviceType});
+                    synCtx.setProperty(MetricConstants.INBOUND_ENDPOINT_LATENCY_TIMER,
+                            metricReporterInstance.getTimer(MetricConstants.INBOUND_ENDPOINT_LATENCY_SECONDS,
+                                    inboundEndpointMap));
+                    break;
+                case SynapseConstants.FAIL_SAFE_MODE_API:
+                    Map<String, String[]> apiMap = new HashMap();
+                    apiMap.put(MetricConstants.API_LATENCY_SECONDS, new String[]{serviceName, serviceType,
+                            apiInvocationUrl});
+
+                    synCtx.setProperty(MetricConstants.API_LATENCY_TIMER,
+                            metricReporterInstance.getTimer(MetricConstants.API_LATENCY_SECONDS, apiMap));
+                    break;
+                default:
+                    log.error("No proper service type found");
+                    break;
+            }
+        }
+    }
+
+    /**
+     * Stop the timer when the response is leaving the Synapse Engine.
+     *
+     * @param timer The Timer object used to observe latency
+     * @ param synCtx The Synapse Message Context
+     */
+    private void stopTimers(Object timer, MessageContext synCtx) {
+        if (null == synCtx.getProperty(SynapseConstants.IS_ERROR_COUNT_ALREADY_PROCESSED) && (null != timer)) {
+             metricReporterInstance.observeTime(timer);
+        }
+    }
+
+    /**
+     * Increment the request count received by a proxy service.
+     *
+     * @param name The metric name
+     */
+    private void incrementProxyCount(String name) {
+        Map<String, String[]> metricCountMap = new HashMap();
+        metricCountMap.put(MetricConstants.PROXY_REQUEST_COUNT_TOTAL, new String[] {name,
+                SynapseConstants.PROXY_SERVICE_TYPE});
+        metricReporterInstance.incrementCount(MetricConstants.PROXY_REQUEST_COUNT_TOTAL, metricCountMap);
+    }
+
+    /**
+     * Increment the request count received by an api.
+     *
+     * @param name The metric name
```

Comment:
Shall we add a javadoc for `apiInvocationUrl` as well

Reply:
Fixed